src/app/service/timer-service.service.ts

```diff
@@ -1,27 +1,141 @@
 
 export type Timer = {
   label: string | null;
   totalTime: number;
   id: number;
+  timerStream?: Observable<number>;
+  status: BehaviorSubject<Status>;
+  destory: Subject<void>;
+  completed: boolean;
 };
 
 @Injectable({
   providedIn: 'root',
 })
 export class TimerService {
-  private timers$ = new ReplaySubject<Timer>();
+  private timers$ = new BehaviorSubject<Timer[]>([]);
 
-  constructor() {}
+  private tick$: Observable<number> = timer(0, 1000);
 
   getTimers(): Observable<Timer[]> {
-    return this.timers$.pipe(
-      scan((acc: Timer[], val: Timer) => [...acc, val], [])
-    );
+    return this.timers$.asObservable();
   }
 
   addTimer(timer: Timer) {
-    this.timers$.next(timer);
+    const array = this.timers$.value;
+    if (array.every((el) => el.id !== timer.id)) {
+      this.timers$.next([...array, timer]);
+    } else {
+      const index = this.findIndex(timer.id);
+      this.timers$.value[index] = timer;
+    }
+  }
+
+  removeTimer(id: number) {
+    const index = this.findIndex(id);
+    this.reset(index);
+    this.timers$.value.splice(index, 1);
+  }
+
+  updateTimerStatus(id: number, value: boolean) {
+    const index = this.findIndex(id);
+    this.timers$.value[index].status.next({ active: value });
+  }
+
+  getTimerStatus(id: number) {
+    const index = this.findIndex(id);
+    return this.timers$.value[index].status.asObservable();
+  }
+
+  resetTimerStream(id: number) {
+    const timer = this.getTimerById(id);
+    const index = this.findIndex(id);
+    this.reset(index);
+    this.addTimer({
+      ...timer,
+      completed: false,
+      timerStream: undefined,
+      destory: new Subject<void>(),
+      status: new BehaviorSubject<Status>({ active: false }),
+    });
+  }
+
+  setTimerStream(id: number, stream$: Observable<number>) {
+    const index = this.findIndex(id);
+    this.timers$.value[index].timerStream = stream$;
+  }
+
+  getTimerStream(id: number) {
+    const index = this.findIndex(id);
+    return this.timers$.value[index].timerStream as Observable<number>;
+  }
+
+  setCompletedStatus(id: number, value: boolean) {
+    const index = this.findIndex(id);
+    this.timers$.value[index].completed = value;
+  }
+
+  getCompletedStatus(id: number) {
+    const index = this.findIndex(id);
+    return this.timers$.value[index].completed;
+  }
+
+  timerStream(id: number): Observable<number> {
+    const timer = this.getTimerById(id);
+    return this.tick$.pipe(
+      takeUntil(timer.destory),
+      withLatestFrom(timer.status),
+      tap(console.log),
+      filter(([_, status]) => status.active),
+      scan((acc, _) => acc - 1, timer.totalTime),
+      takeWhile(Boolean, true),
+      tap((time) => {
+        if (!time) {
+          this.setCompletedStatus(id, true);
+          this.alarm();
+        }
+      }),
+      shareReplay(1)
+    );
+  }
+
+  globalPause() {
+    this.timers$.value.forEach((timer) => timer.status.next({ active: false }));
```

Comment:
How we can pause all without modifying each of the streams?

Reply:
I can create a BehaviorSubject with the boolean value of global pause, add a filter to the tick$ stream with the latest value from global pause, and then map it back to Observable<number>. I think that should pause all streams.